torch/csrc/jit/tensorexpr/bounds_inference.cpp

```diff
@@ -27,36 +27,39 @@ class BoundsInference : public IRVisitor {
 };
 
 void BoundsInference::visit(const Load* v) {
-  accesses_.push_back({v->buf(), kLoad, v->indices(), v->indices()});
+  accesses_[v->buf()].push_back({kLoad, v->indices(), v->indices()});
 }
 
 void BoundsInference::visit(const FunctionCall* v) {
-  accesses_.push_back(
-      {v->tensor()->func_var(), kLoad, v->params(), v->params()});
+  accesses_[v->tensor()->func_var()].push_back(
+      {kLoad, v->params(), v->params()});
 }
 
 void BoundsInference::visit(const Store* v) {
-  accesses_.push_back({v->buf(), kStore, v->indices(), v->indices()});
+  accesses_[v->buf()].push_back({kStore, v->indices(), v->indices()});
   IRVisitor::visit(v);
 }
 
 void BoundsInference::visit(const For* v) {
   v->body()->accept(this);
-  for (TensorAccessBoundsInfo& access : accesses_) {
-    for (size_t j = 0; j < access.start.size(); j++) {
-      // TODO: This function assumes that all indices grow monotonically and
-      // thus for the loop:
-      //   for i in A..B:
-      //     buf[i] = i
-      // the range for i is [A, B). It should be generalized to correctly handle
-      // all cases.
-      const Expr* old_start = access.start[j];
-      const Expr* old_stop = access.stop[j];
-      const Expr* new_start = Substitute(old_start, {{v->var(), v->start()}});
-      const Expr* new_stop =
-          Substitute(old_stop, {{v->var(), new Sub(v->stop(), new IntImm(1))}});
-      access.start[j] = IRSimplifier::simplify(new_start);
-      access.stop[j] = IRSimplifier::simplify(new_stop);
+  for (auto& pair : accesses_) {
+    for (TensorAccessBoundsInfo& access : pair.second) {
```

Comment:
The main reason the original code used a vector instead of a map is to prevent non-determinism (the order in vector was defined by the order of appearance in the statement). I don't think we're using that order anywhere right now, but if we replace vector with a map we should be on a watch out for changes that could accidentally iterate over the map and depend on the order (e.g. if we create some temporary variables as we see entries in the map, it would be quite annoying to see difference in debug dumps from run to run).

Reply:
The map is keyed with the Buf, the accesses themselves are still ordered in the vector (although merging always changes that). Do you really see a situation where where we do rely on the order buffers appear in this list?